ytt-003 Generate OpenAPI Schema

cari-lynn opened the pull request.

Commit:
Create proposal for openapi schema

joe-kimmel-vmw commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,80 @@
+Configuration Authors want to be able to generate documentation for configuration inputs (for example, Data Values Schema, and Data Values) for their users. Additionally, Configuration Consumers want to be able to validate their configuration inputs by other tools (for example, IDE's and OpenAPI Schema validators). 
+
+# Proposal
+Implement a flag to export an [OpenAPI Schema](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) yaml document generated from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file, and an optional Data Values file. When invoking ytt with configuration inputs and the flag `--schema-inspect-openapi`, the result is the OpenAPI formatted document ...
```
feel free to tell me why this is a (much) worse idea:   what if the flag was more like a `-o openapi` ?

cppforlife replied:
-o within ytt controls output format of the data (json, yaml, pos) on top of whats being returned (ie it doesnt control what to return). i dont remember if @cari-lynn and i discussed this variation `ytt -f ... --schema-inspect -o openapi`, but we should give it another thought (perhaps even with -o openapi-short that kills of header or -o openapi for short, and -o openapi-full as long format).

cari-lynn replied:
ytt's `-o` flag right now supports yaml, json, positions. 

The `-o` flag normally applies to the final result of ytt (like Data Values Schema + Data Values + _templates_). For this feature, we want to stop earlier than the final templated result, and only use the configuration inputs, before they are used in templates. Like inspecting the input values after they are all merged together but before they are used, and we have a command that does something like that now: `--data-values-inspect`. So this flag is like that one. It's also not called out here but I would say its even a possibility that the `--schema-inspect-openapi` and `-o json` could be used together at some point.

cppforlife replied:
`-o` flag applies to all output that ytt produces even if we stop within the pipeline midway e.g.

```
$ ytt -v foo=bar --data-values-inspect -ojson
{"foo":"bar"}
```

cari-lynn replied:
As an example, `-o` flag in kubectl allows `json|yaml|name|go-template|go-template-file|template|templatefile|jsonpath|jsonpath-as-json|jsonpath-file`. These seem to me as _how_ should the data be presented, rather than _what_ data to see. I also think it could get complicated when output may be both of an option to `-o` at the same time. For example `ytt -f . -ojson -o openapi`. I updated the proposal to purpose two flags instead `--schema-inspect-openapi` and `--schema-inspect-openapi-metadata`, but happy to hear more thoughts on `-o`.

joe-kimmel-vmw replied:
thanks yall,  I think we can let this thread rest in peace though feel free to reopen if i'm mistaken

cppforlife replied:
im still on it...

> These seem to me as how should the data be presented, rather than what data to see.

exactly. openapi is "how" in my mind at this point. it's a representation of ytt schema in a particular format. this is inline with our existing -opos which shows yaml+positions read of the nodes as its own custom format. now does it make sense to present other non-schema data in openapi format, probably not, but that doesnt mean that it shouldnt be a supported format for schema. with that in mind:

```
ytt --files-inspect                 # prints involved files
ytt --data-values-inspect           # prints values files
ytt --data-values-schema-inspect    # prints ytt schema (as ytt schema with annotations)

ytt --data-values-inspect -ojson                # prints values as json
ytt --data-values-schema-inspect -o openapi     # prints data values schema in openapi format
ytt --data-values-schema-inspect -o openapi-full     # prints data values schema in openapi format with headers
```

cppforlife replied:
also note that i used `--data-values-schema` instead of `--schema` since we are specifically identifying schema for data values `@data/values-schema` vs some generic schema.

pivotaljohn replied:
> openapi is "how" in my mind at this point

I don't think we want the output of `--data-values-schema-inspect` to be limited by what can be expressed in OpenAPI Schema (e.g. certain kinds of validations).
If that's true, then indicating "OpenAPI" _does_ indicate "what" is being output; it's possibly a subset of what would be output for ytt schema.

It also seems reasonable to me that the user may want the OpenAPI Schema content to be in JSON sometimes and YAML other times. So, these two variables seem naturally independent.

If all that resonates, then there are three different data being output: ytt schema, OpenAPI document, and OpenAPI schema component... and it would then fit in the existing scheme to provide a different `inspect` flag for each:

```
ytt --data-values-schema-inspect                 # prints ytt schema for data values
ytt --data-values-schema-inspect-openapi         # prints OpenAPI document for data values
ytt --data-values-schema-inspect-openapi-schema  # prints OpenAPI schema component for data values
```

_(This is essentially an edited set of what @cari-lynn originally proposed.)_

cppforlife replied:
> OpenAPI Schema content to be in JSON sometimes and YAML other times

in my head that would be achieved via stacked output formats e.g. -o openapi,json (or non-cobrafied version like -o openapi -o json). this is pretty common in k8s (-o wide,json)

> I don't think we want the output of --data-values-schema-inspect to be limited by what can be expressed in OpenAPI Schema (e.g. certain kinds of validations).

not sure i understand, by default it would be in native ytt schema format. if you opt into -o openapi, native ytt schema format is represented as openapi. openapi could be seen as an encoding format like base64 or json. in theory we could even decode native ytt schema from openapi format.

cari-lynn replied:
`-o` is growing on me. While it is true that the OpenAPI Schema will be technically a subset of ytt Schema (since ytt Schema in the future may support validations in Starlark) it seems like those differences are limitation of what the OpenAPI spec can represent, rather than an entirely different definition than a ytt Schema. 

If we think of it like a schema is a way to define default values, structure, and types (and eventually validations and documentation..), then both ytt Schema and OpenAPI Schema do those things, just in a different format. In that case the `-o openapi` flag makes sense to toggle the output format between ytt Schema and OpenAPI.
```
ytt --data-values-schema-inspect -o openapi     # prints data values schema in openapi format
ytt --data-values-schema-inspect -o openapi-full     # prints data values schema in openapi format with headers
```

However, we don't currently have a `--data-values-schema-inspect` flag. What would this flag show without the `-o openapi`? The `--data-values-inspect` flag already shows the configuration inputs merged (including the ytt Schema). Does it make sense to require that `--data-values-schema-inspect` always be used with `-o openapi`?

cari-lynn replied:
Since the proposal has been updated since this thread, I'm going to resolve it. Feel free to open it up again or start a new one.

joe-kimmel-vmw commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,80 @@
+openapi: 3.1.0 # Version of the openapi spec
+info:
+  version: 1.0.0 # Version of this document
+  title: Openapi schema generated from ytt schema # Title of this document
```
would we want the title to include a variable portion indicating e.g. filename(s)/paths/ ytt-command-line that was used to assemble?   Can the document provide enough breadcrumbs for a reader / naive maintainer to know how it was generated and how to regenerate or update it?

cari-lynn replied:
Yes, this is a good idea. 
At bare minimum a user could name the file in a way that correlates it to the inputs to ytt (`schema.openapi.yml`?). 

But it does become less straightforward to just add the name of the Data Values Schema file to the title because we may use an arbitrary amount of Data Values Schema and Data Values file all merged together to come up with the result..

joe-kimmel-vmw replied:
should/could we add a metadata section that has even something as crude as `generedated-by: ${args}` ? - if we dump all the command line args into a file is there ever a risk that we'd expose a secret or a credential that was passed in as a data value?      I'm still looking for guidance re. making it easy to reproduce / regenerate / maintain,  vs. the tension of complications.

pivotaljohn replied:
> generedated

stealing this verb. 😜

pivotaljohn replied:
> metadata section that has even something as crude as `generedated-by: ${args}` ?

even better, the OpenAPI specification itself allows for ad-hoc extensions, requiring such fields to be prefixed with `x-` ([ref](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#specificationExtensions)).

So, `x-generedated-by: ${args}`

pivotaljohn replied:
> Can the document provide enough breadcrumbs for a reader / naive maintainer to know how it was generated and how to regenerate or update it?

I believe that in most cases, the invocation of `ytt` to generate the schema will be straight-forward:

```console
$ ytt -f config/ --data-values-schema-inspect-openapi
```

Is that true?


On the implementation side, when we're generating the output, things like command-line args and source files are not in hand... so mating up those sources with the output would introduce some complexity.

cari-lynn replied:
The command to generate a schema should be mostly straightforward, just `ytt -f config/ --data-values-schema-inspect` since it does not accept any additional flags right now. And @pivotaljohn is correct, we lose the information of what command was run.

Since an OpenAPI schema is generated from Data Values Schema at a specific point in time (or at a specific sha), it should be included in version control with the rest of the configuration, and running the command with the same config will yield the same OpenAPI Document. I wonder if there is a way to tie the version of the OpenAPI document to a version of the config it was generated from as a way to make it more clear how it was created..

cppforlife commented on proposals/ytt/003-openapi-schemas/README.md:
````diff
@@ -0,0 +1,80 @@
+```
+#### openapi schema:
+```yaml
+...
+components:
+  schemas:
````
since packaging is not expecting such header (and the root document fields) under spec.valuesSchema.openAPIv3 key (similar to CRDs e.g. https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/config/crds.yml#L16), we should figure out how to disable/enable that within ytt.

cari-lynn replied:
I had an assumption that users outside of kapp controller may want the metadata in the header for info like the OpenAPI `version` and `title`, and for using OpenAPI validation tools that require a absolute compliant OpenAPI document, but users may really only need the `components.schemas` section.

A way could be to have a `--schema-inspect-openapi-schema` flag, which only outputs the `components.schemas` section like kapp-controller wants. And then if users need a complete OpenAPI schema with the header metadata, we could use`--schema-inspect-openapi` flag that may be added later. What do you think?

cari-lynn replied:
I updated the proposal to purpose two flags instead `--schema-inspect-openapi` which would only output the `components.schemas` object, and `--schema-inspect-openapi-metadata`, which would output the entire OpenAPI Doc. We can wait to implement the second flag to hear from users if they need it.

aaronshurley replied:
It could be useful to consider other potential consumers of this functionality. I'm wondering if this capability should live in ytt vs downstream (e.g. kapp-controller).

cari-lynn replied:
Consumers of this functionality would use this to document what Data Values inputs a piece of configuration allows, and potentially validate the Data Values against the OpenAPI Spec using a separate tool. Consumers that want to document inputs will put the openapi schema in a place for config consumers to read, for kapp-controller thats in a Kubernetes resource PackageCR, for others it could be on a website, or included with related configuration in a Git repo. 

In those latter cases, having the additional headers makes more sense than the kapp-controller case, since a PackageCR already has the OpenAPI version included, but if it were just a file in a repo that version info would be missing.

cari-lynn replied:
An update to this proposal includes header information in the `ytt -f config/ --data-values-schema-inspect` command. This follows the conventions other tools use (for example kubectl get --raw openapi/v2) and also makes the Document fully compliant with the OpenAPI Spec. There is another flag `--data-values-schema-inspect -o openapi-schema` which would output only the `components.schemas` section.

Commit:
Update proposal add flag and output schema object only

- The `--schema-inspect-openapi` flag only includes a subsection of
  openapi document
- Add suggestion for new flag to output the entire openapi docment`

danielhelfand reviewed (commented):
A couple comments and a question:

What would be our recommended worfklow for including this information in a kapp-controller Package CR without having to manually copy contents of generated document to the Package CR valuesSchema property?

Maybe we consider this out of scope for now, which I think is totally fair, but curious if we are thinking of including a way to patch an existing Package CR with this generated schema?

More info on Package CR/valuesSchema here: https://carvel.dev/kapp-controller/docs/latest/packaging/#package-1

aaronshurley commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,85 @@
+Configuration Authors want to be able to generate documentation for configuration inputs (for example, Data Values Schema, and Data Values) for their users. Additionally, Configuration Consumers want to be able to validate their configuration inputs by other tools (for example, IDE's and OpenAPI Schema validators). 
+
+# Proposal
+Implement a flag to create an [OpenAPI Schema](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) yaml document generated from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file, and an optional Data Values file. When invoking ytt with configuration inputs and the flag `--schema-inspect-openapi`, the result is the OpenAPI formatted Schema o ...
```
My guess is that `--schema-inspect-openapi` provides some consistency with some existing flags such as `--data-values-inspect` and `--files-inspect`. However, I do see some differences between the functionality of the proposed vs existing flags. The existing flags are providing information to the user about what ytt is processing (i.e. inspecting). The new flags create an OpenAPI schema based on inputs provided to ytt (this feels like it's beyond "just inspecting"). I'd agree with considering a name change to something like `--schema-create-openapi` or `--schema-generate-openapi` (or `--schema-openapi-create`).

pivotaljohn replied:
That's an interesting point, @danielhelfand. There's a subtlety here in that the OpenAPI Schema component will likely always be a subset of what the `ytt` schema will be. Using a verb that contains a meaning of "producing" (like those suggested, so far), captures that distinction.

More explicitly:

`--data-values-schema-inspect` would output _all_ schema information (as a `ytt` schema document).
whereas, `--data-values-schema-generate-openapi` would produce an OpenAPI document _based on_ the schema calculated in the execution.

cari-lynn replied:
We touched on this topic in the Carvel Community Meeting Thurs 8/19, and it was mentioned that this flag is very similar to ones that exist:
```bash
# proposed flag
ytt --data-values-schema-inspect -o openapi # prints data values schema files as an OpenAPI Document

# current flags
ytt --files-inspect                 # prints involved files
ytt --data-values-inspect           # prints values files
ytt --data-values-inspect -ojson    # prints values as json
```
It takes an input of Data Values Schema files, and merges all the files together, then outputs that information in an OpenAPI way, and could support other formats as well, such as ytt Schema format, or JsonSchema in the future. My point being that it isn't exactly generating or creating something brand new per se, but rather outputting an input to ytt in a specific way.

aaronshurley commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,85 @@
+
+# Proposal
+Implement a flag to create an [OpenAPI Schema](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) yaml document generated from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file, and an optional Data Values file. When invoking ytt with configuration inputs and the flag `--schema-inspect-openapi`, the result is the OpenAPI formatted Schema o ...
+In addition, the `--schema-inspect-openapi-metadata` flag may be used to include _both_ the [OpenAPI metadata](#openapi-document-with-metadata) and the [Schema object](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object).
```
@danielhelfand Are you suggesting that (a) the OpenAPI metadata is unnecessary, (b) that the OpenAPI metadata should always be returned, or (c) something else?

danielhelfand replied:
I am referring to just the name of the flag (i.e. `--schema-inspect-openapi-metadata`). It's already a lengthy name and to me the word metadata does not make the purpose much clearer. I think `--schema-inspect-openapi` would be clearer, but I am guessing metadata was added since the generation flag is named `--schema-inspect-openapi`. My [comment](https://github.com/vmware-tanzu/carvel/pull/198#discussion_r685266514) above mentions that I think that flag name should be something different.

pivotaljohn replied:
In fact, when I see the name `--...openapi-metadata`, I would presume that it would include _only_ the metadata (i.e. API name and Info object).

When I move my frame of reference over to the OpenAPI world and look back at `ytt`, then I would assume that the default output is a proper/complete OpenAPI Document (which is exactly what's in the proposal). And from that same perspective, limiting the output to the Schema component feels like a flavor-of the OpenAPI Document.

A corresponding (evolved) set of flags is detailed in [this other comment](https://github.com/vmware-tanzu/carvel/pull/198#discussion_r687197686).

cari-lynn replied:
Great points! ❤️  The flag name has been changed in the updated proposal.

pivotaljohn commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,85 @@
+### Scenario 3: Data Values file only
+`ytt -f values.yml --schema-inspect-openapi` will result in a warning suggesting that a Data Values Schema file be provided. 
+
+The reason for waiting to support Data Values files only is so that the OpenAPI document can have the additional information of default values and descriptions only available in Data Values Schema files. If this functionality comes with no additional work, we should reevaluate this decision.
```
I personally have been viewing using a Data Values Overlay to provide defaults for an array type as a workaround rather than _the_ way.

If we had already implemented [`@schema/default`](https://hackmd.io/pODV3wzbT56MbQTxbQOOKQ?view#schemadefault) would it obviate the desire to include data values in this output?
If so, I wonder if we want to defer supporting non-empty default values in "schema inspection"... then, when we _do_ implement the ability to explicitly set a default in schema, it all "just works."

There's a little bit of implementation poking up here: schema is calculated in a completely separate phase _before_ data values are. If we design-in including data values in this inspection, it would need to occur long after schema is actually calculated.
Perhaps this is overly pedantic thinking, but shouldn't we consider schema as separate from data values?

cari-lynn replied:
The main motivator for using the merged Data Values Schema _and_ Data Values in the output is so that a configuration _consumer_ can know what default values they need to override. For example, if there exists
```
#@data/values-schema 
maximum: 1
```
```
#@data/values 
maximum: 2
```
And we only use the Data Values Schema file and not the Data Values file, then the OpenAPI schema would be:
```
maximum:
  type: integer
  default: 1
```
In the case of kapp-controller, the entire configuration may not always be available to the configuration consumer, if they only read this OpenAPI Schema they would have the impression the value for `maximum` will be 1, and if that is the value they need, then they would not modify or override it.

However when invoking `ytt -f schema.yml -f values.yml` the value for `maximum` would actually be 2. Which would not be what they would expect.

pivotaljohn replied:
Thank you for explaining that. Yeah, I see how what's relevant to me as a Configuration Consumer is the _net_ default values. 👍🏻

cari-lynn replied:
Update: While the point above about configuration consumers need to know what default values they need to override still holds true, the proposal has been updated to only include Data Values Schema files in the `--data-values-schema-inspect` command. 

> but shouldn't we consider schema as separate from data values?

Yes. Data Values Schema files are able to accomplish everything that a Data Values file can do (caveat - this will only be true once a bit more issues have been implemented), and for most cases configuration authors can use Data Values Schema files for everything they want to do with Data Values files. This should eliminate the need for Data Values overlays. 

By allowing use of only Data Value Schema files for now, we are able to take advantage of the additional information they provide that is helpful in an OpenAPI Doc. If the need for supporting Data Values overlays does present itself in the future, we can implement this flag: `ytt -f config/ --data-values-inspect -o oepnapi` which will support it.

> If we had already implemented @schema/default would it obviate the desire to include data values in this output?

As part of this work we will need to implement `#@schema/default in order to only use Data Values Schema files.

pivotaljohn commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,85 @@
+      type: boolean
+    static_ip: 
+      type: string
+  required:
+  - enabled
+  - static_ip
```
How would one define a field in ytt schema that would be considered "not required" in OpenAPI schema?

Commit:
Add Use Cases and change flag name

cari-lynn commented:
I made some changes to the proposal based on feedback (thank you!)
- I changed the flag name to `--data-values-schema-inspect` that can be used with `-o openapi`, `-o openapi-full`
-  I added Use Cases for kapp-controller to address one of your comments @danielhelfand 

I will be addressing more feedback soon thank you for your patience

Commit:
Major Update include headers in output

- Use only Data Values Schemas for `--data-value-schema-inspect`
- Use complete OpenAPI Document rather than Schema only

cari-lynn commented:
Once again the proposal has been updated to reflect discussions had this week. Notably:
- Use full OpenAPI Document with headers and Schema 
  - This aligns with how other tools use OpenAPI, such as `kubectl get --raw /openapi/v2`
- Distinct flags for an OpenAPI Document created from just a Data Values Schema and one that also includes a Data Values file.

Please feel free to add feedback on this.

pivotaljohn reviewed (changes requested):
Overall converging. 👍🏻 

Looking for a few clarifications.

cari-lynn committed:
Add yaml format to flag

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,148 @@
+
+The `--data-values-inspect -o openapi` flag creates an OpenAPI Document from a Data Values Schema, and a Data Values file. The flag `--data-values-inspect` currently exists and outputs the values that result from merging Data Values Schema and Data Values together. The addition of `-o openapi` would use the merged values of Data Values Schema and Data Values to output an OpenAPI Document.
+
+The  `--data-values-schema-inspect` flag creates a ytt schema with annotations (_not in OpenAPI format_) from a Data Values Schema file.
```
Yes, the behavior should be that prior to that format being implemented, when using `--data-value-schema-inspect` without `-o openapi` Cobra should report an error saying `-o` is required. For example: `The following flags are required and were not present: "--output"`

Commit:
Wording on merging 1+ Data Values Schemas

cari-lynn commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,148 @@
+ytt --data-values-inspect -ojson    # prints values as json
+
+# New flags
+ytt --data-values-schema-inspect -o openapi     # prints data values schema files as an OpenAPI Document
```
Updated: I called out that the result is one OpenAPI Document created from 1+ merged Data Values Schema

cari-lynn commented on proposals/ytt/003-openapi-schemas/README.md:
````diff
@@ -0,0 +1,148 @@
+1. Create an OpenAPI Document from a Data Values Schema file: `ytt -f schema.yml --data-values-schema-inspect -o openapi`.
+```yaml
+#! openapi.yml
+openapi: 3.1.0
````
I did some [research](https://gist.github.com/cari-lynn/bc35270734e41ea17947e2f43a631494) on the differences between 3.0.x and 3.1.0. They are very differencet specs, with breaking changes. 3.1.0 was released this year, and has full compatability with JsonSchema and some nice advantages. However 3.1.0 does not have very much adoption based on [this site](https://openapi.tools/) and a look through the GitHub OpenAPI 3 topic. Because it hasn't been adopted by many tools yet if users want to use a tool to validate their OpenAPI Document they may not be able to find one yet. For this reason I think we should use 3.0.x.

cari-lynn commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,148 @@
+Configuration Authors want to be able to generate documentation for configuration inputs (for example, Data Values Schema, and Data Values) for their users. Additionally, Configuration Consumers want to be able to validate their configuration inputs by other tools (for example, IDE's and OpenAPI Schema validators). 
+
+# Proposal
+Implement the flag `--data-values-schema-inspect -o openapi` to create an [OpenAPI Document](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#openapi-object) from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file. When invoking ytt with this flag and a Data Values Schema file, a yaml OpenAPI document that has [ headers](#openapi-document- ...
```
I think it makes sense to use `-o openapi-v3` as the flag name. It will allow us to support any number of versions. For example:  `-o openapi-v3.1`.

Commit:
Use OpenAPI 3.0.3 and modify flag name

Commit:
Add Implementation phases and details

pivotaljohn reviewed (approved):
Excellent work, @cari-lynn.

This reads like an (even more) intelligent start to the track to me. 👍🏻

cari-lynn commented on proposals/ytt/003-openapi-schemas/README.md:
```diff
@@ -0,0 +1,191 @@
+
+The `#@schema/doc` annotation will need to be implemented to provide the `description` key that explains what a key in a schema is.
+
+The `#@schema/example` annotation will need to be implemented to provide an `example` key for a key in the schema.
```
The OpenAPI 3.0 spec we are planning to use only supports a [single example](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#fixed-fields-20). So we could implement `#@schema/examples` and only use the first example, or we can just implement `#@schema/example`. I think the second option would be more clear what would be exported.

cari-lynn commented:
This proposal has been accepted, and implementation work has begun. I will leave this PR open until Friday September 17th, so that any questions, comments, feedback can be discussed here prior to the 17th.
